Task/web apps build tree

edsanville opened the pull request:
* Web build doesn't touch the src tree (except for src/web/node_modules, because ts-loader cannot access the modules unless it's there... lame!)
* Got rid of .gitignores for artifacts
* Silenced some warnings, and silenced webpack and npm unless there are errors

edsanville committed:
Remove .gitignore entries for build artifacts

edsanville committed:
Fix install_dependencies.sh and build_messages.sh to target a specified directory

edsanville committed:
Build JCC directly into a target directory

edsanville committed:
TypeScript outputs into intermediate build directory

edsanville committed:
The run.sh script serves web apps from the BUILD_DIR

edsanville committed:
run.sh symlinks jed to the build dir, so we can hot reload while developing

edsanville committed:
Add build_venv.sh script

edsanville committed:
Build jdv into a target directory

edsanville committed:
Disable flask DEBUG for jdv
Use the test.webpack.config.js for local testing of jdv

edsanville committed:
Use build directory targeting scripts for web CMake build

edsanville committed:
Silence jdv build stats

edsanville committed:
Remove extraneous proto imports

edsanville committed:
Silence the global bind placeholder warning

edsanville committed:
Remove extraneous gitignore files

edsanville committed:
Remove additional extraneous proto imports

edsanville committed:
Test in release mode, with watch enabled

edsanville committed:
Build with webpack errors-only during release build

edsanville committed:
We want to print errors

edsanville committed:
install_dependencies.sh should use its command-line parameter
npm should be quiet unless there's an error

edsanville committed:
Get rid of npm scripts for building jcc

tsaubergine commented:
Is there *really no way* to move node_modules and package-lock.json out of the source tree?

tsaubergine commented on src/web/CMakeLists.txt:
```diff
@@ -28,7 +28,9 @@ set(command_control_outdir ${project_SHARE_DIR}/jaiabot/web/command_control/dist
   ARGS ${command_control_outdir}
+  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/command_control
   COMMENT "Running command_control build.sh"
+  DEPENDS npm_depdendencies
```
Typo - I realize this doesn't affect functionality but could lead to subtle errors in the future and looks better for anyone looking at the code.

tsaubergine replied:
Also I'm pretty sure all the *.js, *.ts, etc. files (anything that webpack uses as source) need to be listed here as dependencies so this is re-run by CMake if those are changed.

tsaubergine reviewed (commented):
I'd like to see the directory structure consistent between jcc/jdv as well:

Perhaps:
web/command_control -> web/jcc
web/server -> web/jcc/server

I don't think we're using jaia_cockpit_extension? If not this can go away and we can resurrect it from the history if needed.

tsaubergine commented on src/web/install_dependencies.sh:
```diff
@@ -1,45 +1,11 @@
+PACKAGE_JSON_DIR=$1
 
-echo 🟢 Installing npm dependencies to: $NODE_MODULES_DIR
+echo 🟢 Installing npm dependencies in $(pwd) \(sorry, ts-loader needs them there\)
```
`$(pwd)` is the current working directory, not where the script is located (which is what I think you want). `$(dirname "$0")` is usually sufficient for that.

edsanville replied:
Ah yes, this comment is actually obsolete and should be reverted because it's actually being passed the path via argv[1] again.  Before I was only calling it from the pwd, so I used that, but I like the simple way of getting the script's directory.

This should all be obsolete when it's incorporated into the CMakeLists

tsaubergine commented on src/web/jdv/CMakeLists.txt:
```diff
@@ -2,12 +2,11 @@ set(jdv_outdir ${project_SHARE_DIR}/jaiabot/web/jdv/)
   ARGS ${jdv_client_outdir}
+  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/client
   COMMENT "Running jdv build.sh"
   DEPENDS npm_depdendencies
```
Same comment as with JCC: the DEPENDS should list all source files for webpack

edsanville replied:
Ah yes, I think one of the next cards involves making sure the DEPENDS all include everything they really DEPEND on.

tsaubergine commented on src/web/jdv/client/build.sh:
Can we do away with these little build.sh scripts in favor of just using CMake?

edsanville replied:
Yes, I think that makes sense.

edsanville replied:
I think we should keep some of these scripts, like build_venv.sh, because they're useful for the run.sh scripts that we use during web development.  We don't have to use them in the CMake build, though.

tsaubergine commented:
I'm getting this error on Ubuntu 22.04: 

```
node:internal/modules/cjs/loader:1051
  throw err;
  ^

Error: Cannot find module 'resolve-cwd'
Require stack:
- /usr/share/nodejs/webpack/node_modules/import-local/index.js
- /usr/share/nodejs/webpack/node_modules/webpack-cli/bin/cli.js
- /usr/share/nodejs/webpack/bin/webpack.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
    at Module._load (node:internal/modules/cjs/loader:901:27)
    at Module.require (node:internal/modules/cjs/loader:1115:19)
    at require (node:internal/modules/helpers:119:18)
    at Object.<anonymous> (/usr/share/nodejs/webpack/node_modules/import-local/index.js:3:20)
    at Module._compile (node:internal/modules/cjs/loader:1233:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1287:10)
    at Module.load (node:internal/modules/cjs/loader:1091:32)
    at Module._load (node:internal/modules/cjs/loader:938:12)
    at Module.require (node:internal/modules/cjs/loader:1115:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/usr/share/nodejs/webpack/node_modules/import-local/index.js',
    '/usr/share/nodejs/webpack/node_modules/webpack-cli/bin/cli.js',
    '/usr/share/nodejs/webpack/bin/webpack.js'
  ]
}

Node.js v20.5.0
node:internal/modules/cjs/loader:1051
  throw err;
  ^

Error: Cannot find module 'resolve-cwd'
Require stack:
- /usr/share/nodejs/webpack/node_modules/import-local/index.js
- /usr/share/nodejs/webpack/node_modules/webpack-cli/bin/cli.js
- /usr/share/nodejs/webpack/bin/webpack.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
    at Module._load (node:internal/modules/cjs/loader:901:27)
    at Module.require (node:internal/modules/cjs/loader:1115:19)
    at require (node:internal/modules/helpers:119:18)
    at Object.<anonymous> (/usr/share/nodejs/webpack/node_modules/import-local/index.js:3:20)
    at Module._compile (node:internal/modules/cjs/loader:1233:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1287:10)
    at Module.load (node:internal/modules/cjs/loader:1091:32)
    at Module._load (node:internal/modules/cjs/loader:938:12)
    at Module.require (node:internal/modules/cjs/loader:1115:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/usr/share/nodejs/webpack/node_modules/import-local/index.js',
    '/usr/share/nodejs/webpack/node_modules/webpack-cli/bin/cli.js',
    '/usr/share/nodejs/webpack/bin/webpack.js'
  ]
}

Node.js v20.5.0
make[2]: *** [src/web/CMakeFiles/command_control.dir/build.make:76: share/jaiabot/web/command_control/dist/client/client.js] Error 1
make[1]: *** [CMakeFiles/Makefile2:1516: src/web/CMakeFiles/command_control.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make[2]: *** [src/web/jdv/CMakeFiles/jdv.dir/build.make:73: share/jaiabot/web/jdv/client/dist/bundle.js] Error 1
make[1]: *** [CMakeFiles/Makefile2:1542: src/web/jdv/CMakeFiles/jdv.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
Any thoughts?

edsanville commented:
> Is there _really no way_ to move node_modules and package-lock.json out of the source tree?

Maybe there is, but I spent well over an hour trying to accomplish it to no avail.  The webpack program can build against a node_modules in any arbitrary location, but there was still an issue compiling the TypeScript files.  I searched and searched for a way to get ts-loader to import against the correct node_modules, but I gave up and accepted that it couldn't be done in a sane way.

There's a caveat to this:  we could possibly install the node_modules in the parent directory of the jaiabot directory, since webpack and ts-loader both use the algorithm of climbing the directories until they find a node_modules.  It seems pretty radical to me, but I'm up for it if we agree on that.

These web toolchains are much more obnoxious to use than the Cxx ones, as I'm sure you've noticed.  It's hard to get them to shut up printing to stdout/err, and it's hard to point them to arbitrary directories.

edsanville commented:
> > Is there _really no way_ to move node_modules and package-lock.json out of the source tree?
> 
> Maybe there is, but I spent well over an hour trying to accomplish it to no avail. The webpack program can build against a node_modules in any arbitrary location, but there was still an issue compiling the TypeScript files. I searched and searched for a way to get ts-loader to import against the correct node_modules, but I gave up and accepted that it couldn't be done in a sane way.
> 
> There's a caveat to this: we could possibly install the node_modules in the parent directory of the jaiabot directory, since webpack and ts-loader both use the algorithm of climbing the directories until they find a node_modules. It seems pretty radical to me, but I'm up for it if we agree on that.
> 
> These web toolchains are much more obnoxious to use than the Cxx ones, as I'm sure you've noticed. It's hard to get them to shut up printing to stdout/err, and it's hard to point them to arbitrary directories.

I should also comment though, that installing node_modules in the parent of jaiabot would kind of defeat the biggest purpose (IMO), which is that all our build products are in jaiabot/build.

tsaubergine commented:
> > Is there _really no way_ to move node_modules and package-lock.json out of the source tree?
> 
> Maybe there is, but I spent well over an hour trying to accomplish it to no avail. The webpack program can build against a node_modules in any arbitrary location, but there was still an issue compiling the TypeScript files. I searched and searched for a way to get ts-loader to import against the correct node_modules, but I gave up and accepted that it couldn't be done in a sane way.
> 
> There's a caveat to this: we could possibly install the node_modules in the parent directory of the jaiabot directory, since webpack and ts-loader both use the algorithm of climbing the directories until they find a node_modules. It seems pretty radical to me, but I'm up for it if we agree on that.
> 
> These web toolchains are much more obnoxious to use than the Cxx ones, as I'm sure you've noticed. It's hard to get them to shut up printing to stdout/err, and it's hard to point them to arbitrary directories.

What about using cmake `configure_file` to copy the source files into build before running webpack?

edsanville commented:
> I'd like to see the directory structure consistent between jcc/jdv as well:
> 
> Perhaps: web/command_control -> web/jcc web/server -> web/jcc/server
> 
> I don't think we're using jaia_cockpit_extension? If not this can go away and we can resurrect it from the history if needed.

I thought the same thing about the directory structure...  we can add it to the list of tasks.

edsanville commented:
> > > Is there _really no way_ to move node_modules and package-lock.json out of the source tree?
> > 
> > 
> > Maybe there is, but I spent well over an hour trying to accomplish it to no avail. The webpack program can build against a node_modules in any arbitrary location, but there was still an issue compiling the TypeScript files. I searched and searched for a way to get ts-loader to import against the correct node_modules, but I gave up and accepted that it couldn't be done in a sane way.
> > There's a caveat to this: we could possibly install the node_modules in the parent directory of the jaiabot directory, since webpack and ts-loader both use the algorithm of climbing the directories until they find a node_modules. It seems pretty radical to me, but I'm up for it if we agree on that.
> > These web toolchains are much more obnoxious to use than the Cxx ones, as I'm sure you've noticed. It's hard to get them to shut up printing to stdout/err, and it's hard to point them to arbitrary directories.
> 
> What about using cmake `configure_file` to copy the source files into build before running webpack?

Interesting...  this could work, I think.  Are you thinking copy everything to build/intermediate, and then build it there, targeting the final directory?

tsaubergine commented:
> > > > Is there _really no way_ to move node_modules and package-lock.json out of the source tree?
> > > 
> > > 
> > > Maybe there is, but I spent well over an hour trying to accomplish it to no avail. The webpack program can build against a node_modules in any arbitrary location, but there was still an issue compiling the TypeScript files. I searched and searched for a way to get ts-loader to import against the correct node_modules, but I gave up and accepted that it couldn't be done in a sane way.
> > > There's a caveat to this: we could possibly install the node_modules in the parent directory of the jaiabot directory, since webpack and ts-loader both use the algorithm of climbing the directories until they find a node_modules. It seems pretty radical to me, but I'm up for it if we agree on that.
> > > These web toolchains are much more obnoxious to use than the Cxx ones, as I'm sure you've noticed. It's hard to get them to shut up printing to stdout/err, and it's hard to point them to arbitrary directories.
> > 
> > 
> > What about using cmake `configure_file` to copy the source files into build before running webpack?
> 
> Interesting... this could work, I think. Are you thinking copy everything to build/intermediate, and then build it there, targeting the final directory?

Yes my thought is to have CMake build a list of the dependencies (source files), configure_file( COPYONLY) these, and then build (and if you use the same list of dependencies to pass to add_custom_command as well you're sure you're not missing anything).

This is similar to how I create build/include directories from the .h files in the src tree.

edsanville commented:
> > > > > Is there _really no way_ to move node_modules and package-lock.json out of the source tree?
> > > > 
> > > > 
> > > > Maybe there is, but I spent well over an hour trying to accomplish it to no avail. The webpack program can build against a node_modules in any arbitrary location, but there was still an issue compiling the TypeScript files. I searched and searched for a way to get ts-loader to import against the correct node_modules, but I gave up and accepted that it couldn't be done in a sane way.
> > > > There's a caveat to this: we could possibly install the node_modules in the parent directory of the jaiabot directory, since webpack and ts-loader both use the algorithm of climbing the directories until they find a node_modules. It seems pretty radical to me, but I'm up for it if we agree on that.
> > > > These web toolchains are much more obnoxious to use than the Cxx ones, as I'm sure you've noticed. It's hard to get them to shut up printing to stdout/err, and it's hard to point them to arbitrary directories.
> > > 
> > > 
> > > What about using cmake `configure_file` to copy the source files into build before running webpack?
> > 
> > 
> > Interesting... this could work, I think. Are you thinking copy everything to build/intermediate, and then build it there, targeting the final directory?
> 
> Yes my thought is to have CMake build a list of the dependencies (source files), configure_file( COPYONLY) these, and then build (and if you use the same list of dependencies to pass to add_custom_command as well you're sure you're not missing anything).
> 
> This is similar to how I create build/include directories from the .h files in the src tree.

Great, that sounds like an optimal solution.

edsanville committed:
Copy web src trees into build/<arch>/intermediate/web, and then install node_modules there, and build apps into their final locations
Build process doesn't generate any node_modules directory anywhere in the src tree anymore
Renamed command_control to jcc

edsanville committed:
Remove the node_modules and package-lock.json ignores

edsanville committed:
Remove jaia_cockpit_extension (for now)

edsanville committed:
Move server directory under jcc
Remove build.sh scripts

edsanville commented:
> I'm getting this error on Ubuntu 22.04:
> 
> ```
> node:internal/modules/cjs/loader:1051
>   throw err;
>   ^
> 
> Error: Cannot find module 'resolve-cwd'
> Require stack:
> - /usr/share/nodejs/webpack/node_modules/import-local/index.js
> - /usr/share/nodejs/webpack/node_modules/webpack-cli/bin/cli.js
> - /usr/share/nodejs/webpack/bin/webpack.js
>     at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
>     at Module._load (node:internal/modules/cjs/loader:901:27)
>     at Module.require (node:internal/modules/cjs/loader:1115:19)
>     at require (node:internal/modules/helpers:119:18)
>     at Object.<anonymous> (/usr/share/nodejs/webpack/node_modules/import-local/index.js:3:20)
>     at Module._compile (node:internal/modules/cjs/loader:1233:14)
>     at Module._extensions..js (node:internal/modules/cjs/loader:1287:10)
>     at Module.load (node:internal/modules/cjs/loader:1091:32)
>     at Module._load (node:internal/modules/cjs/loader:938:12)
>     at Module.require (node:internal/modules/cjs/loader:1115:19) {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: [
>     '/usr/share/nodejs/webpack/node_modules/import-local/index.js',
>     '/usr/share/nodejs/webpack/node_modules/webpack-cli/bin/cli.js',
>     '/usr/share/nodejs/webpack/bin/webpack.js'
>   ]
> }
> 
> Node.js v20.5.0
> node:internal/modules/cjs/loader:1051
>   throw err;
>   ^
> 
> Error: Cannot find module 'resolve-cwd'
> Require stack:
> - /usr/share/nodejs/webpack/node_modules/import-local/index.js
> - /usr/share/nodejs/webpack/node_modules/webpack-cli/bin/cli.js
> - /usr/share/nodejs/webpack/bin/webpack.js
>     at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
>     at Module._load (node:internal/modules/cjs/loader:901:27)
>     at Module.require (node:internal/modules/cjs/loader:1115:19)
>     at require (node:internal/modules/helpers:119:18)
>     at Object.<anonymous> (/usr/share/nodejs/webpack/node_modules/import-local/index.js:3:20)
>     at Module._compile (node:internal/modules/cjs/loader:1233:14)
>     at Module._extensions..js (node:internal/modules/cjs/loader:1287:10)
>     at Module.load (node:internal/modules/cjs/loader:1091:32)
>     at Module._load (node:internal/modules/cjs/loader:938:12)
>     at Module.require (node:internal/modules/cjs/loader:1115:19) {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: [
>     '/usr/share/nodejs/webpack/node_modules/import-local/index.js',
>     '/usr/share/nodejs/webpack/node_modules/webpack-cli/bin/cli.js',
>     '/usr/share/nodejs/webpack/bin/webpack.js'
>   ]
> }
> 
> Node.js v20.5.0
> make[2]: *** [src/web/CMakeFiles/command_control.dir/build.make:76: share/jaiabot/web/command_control/dist/client/client.js] Error 1
> make[1]: *** [CMakeFiles/Makefile2:1516: src/web/CMakeFiles/command_control.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> make[2]: *** [src/web/jdv/CMakeFiles/jdv.dir/build.make:73: share/jaiabot/web/jdv/client/dist/bundle.js] Error 1
> make[1]: *** [CMakeFiles/Makefile2:1542: src/web/jdv/CMakeFiles/jdv.dir/all] Error 2
> make: *** [Makefile:136: all] Error 2
> ```
> 
> Any thoughts?

StackOverflow seems to think this is due to some combination of:
* Old version of webpack being used
* Dirty node_modules directory
* Old version of node being used

It's possible the new setup-tools-build.sh and deploy.sh (which makes sure it's using the nvm-managed versions of the tools), could avoid this error.  Especially when combined with a clean `build/intermediate` directory in the latest commit?
